src/inmanta/server/services/resourceservice.py

```diff
@@ -426,115 +426,125 @@ def in_requires(req: ResourceVersionIdStr) -> bool:
+                resource_action = await data.ResourceAction.get(action_id=action_id, connection=connection)
+                if resource_action is None:
+                    # new
+                    if started is None:
+                        return 500, {"message": "A resource action can only be created with a start datetime."}
+
+                    version = Id.parse_id(resource_ids[0]).version
+                    resource_action = data.ResourceAction(
+                        environment=env.id,
+                        version=version,
+                        resource_version_ids=resource_ids,
+                        action_id=action_id,
+                        action=action,
+                        started=started,
+                    )
+                    await resource_action.insert(connection=connection)
+                else:
+                    # existing
+                    if resource_action.finished is not None:
+                        return (
+                            500,
+                            {
+                                "message": (
+                                    "An resource action can only be updated when it has not been finished yet. This action "
+                                    "finished at %s" % resource_action.finished
+                                )
+                            },
+                        )
+
+                if len(messages) > 0:
+                    resource_action.add_logs(messages)
+                    for msg in messages:
+                        # All other data is stored in the database. The msg was already formatted at the client side.
+                        self.log_resource_action(
+                            env.id,
+                            resource_ids,
+                            const.LogLevel[msg["level"]].value,
+                            datetime.datetime.strptime(msg["timestamp"], const.TIME_ISOFMT),
+                            msg["msg"],
+                        )
+
+                if len(changes) > 0:
+                    resource_action.add_changes(changes)
+
+                if status is not None:
+                    resource_action.set_field("status", status)
+
+                if change is not None:
+                    resource_action.set_field("change", change)
+
+                resource_action.set_field("send_event", send_events)
+
+                if finished is not None:
+                    resource_action.set_field("finished", finished)
+
+                await resource_action.save(connection=connection)
+
+                if is_resource_state_update:
+                    # transient resource update
+                    if not is_resource_action_finished:
+                        for res in resources:
+                            await res.update_fields(status=status, connection=connection)
+                        if not keep_increment_cache:
+                            self.clear_env_cache(env)
```

Comment:
I created a separate issue to look into the env_cache 
https://github.com/inmanta/inmanta/issues/2022